Upgrading the documentation of neovim

SeniorMars opened the pull request:
I aim to have Neovim's documentation similar to Vim's.

Added examples to:
- [x] options.c (nvim_{s,g}et_option{_value})
- [x] buffer.c (nvim_buf_*)
- [x] window.c (nvim_win_*)
- [x] win_config.c
- [x] autocmd.c
- [x] tabpage.c
- [ ] command.c

SeniorMars committed:
docs: [WIP] improving documentation

docs: Improved readability of api.text

This is the first of many commits, where I will be tackling on
documentation in Neovim.

In this commit I
- Grammar checked api.txt through LanguageTool.
- Tried to format lines to be under 80 chars.
- Changed the documentation for some functions.
- Fixed many spelling mistakes.

My goal is to have Neovim's documentation on a similar level as Vim's.

docs: Updated options.c documentation with examples

In addition to grammar, I also added examples I thought were relevant.

docs: Added spacing and lua tag to examples

docs: Added @note where it is needed

Added examples to nearly all buffer api functions

I tried my best to introduce examples with practical use cases, and I
did this by actively looking at examples from plugins/libraries on
github.

Additionally, I added a few commas..., which can introduce different
pose.

docs: reverted some changes

SeniorMars committed:
Improved readability of api.text

This is the first of many commits, where I will be tackling on
documentation in Neovim.

In this commit I
- Grammar checked api.txt through LanguageTool.
- Tried to format lines to be under 80 chars.
- Changed the documentation for some functions.
- Fixed many spelling mistakes.

My goal is to have Neovim's documentation on a similar level as Vim's.

justinmk commented on runtime/doc/api.txt:
```diff
@@ -179,25 +179,25 @@ Nvim exposes its API metadata as a Dictionary with these items:
      can define functions at runtime.
 
-  2. Start Nvim with |--api-info|. Useful for statically-compiled clients.
+  2. Start Nvim with |--api-info|. Useful for statically compiled clients.
```
?

justinmk commented on runtime/doc/api.txt:
```diff
@@ -646,7 +646,7 @@ nvim_call_atomic({calls})                                 *nvim_call_atomic()*
 
     Return: ~
-        Array of two elements. The first is an array of return values. The
+        An array of two elements. The first is an array of return values. The
```
Do not do add articles everywhere. This is explicitly called out in `:help dev-doc`.

justinmk commented on runtime/doc/api.txt:
```diff
@@ -656,8 +656,9 @@ nvim_call_atomic({calls})                                 *nvim_call_atomic()*
     process. For the stdio channel |channel-stdio|, it writes to Nvim's
-    stdout. For an internal terminal instance (|nvim_open_term()|) it writes
-    directly to terminal output. See |channel-bytes| for more information.
+    stdout. In the case of an internal terminal instance (|nvim_open_term()|)
```
"For => In the case" is a random stylistic change that increases verbosity. Don't want to encourage this.

justinmk commented on runtime/doc/api.txt:
```diff
@@ -1561,7 +1562,7 @@ nvim_command({command})                                       *nvim_command()*
     construct an Ex command using a structured format and then execute it, use
     |nvim_cmd()|. To modify an Ex command before evaluating it, use
-    |nvim_parse_cmd()| in conjunction with |nvim_cmd()|.
+    |nvim_parse_cmd()| with |nvim_cmd()|.
```
This is a good change. Reduces verbosity without losing clarity.

clason reviewed (changes requested):
Most of these are autogenerated, so you need to edit the docstrings in the C source code and run `./scripts/gen_vimdoc.py`; your changes here will be simply overwritten.

Also, please separate simple typo/punctuation changes (which get folded into an automated PR) from actual documentation changes (which need a proper review, as some changes are in fact wrong).

justinmk commented on runtime/doc/api.txt:
```diff
@@ -1726,7 +1727,7 @@ nvim_cmd({*cmd}, {*opts})                                         *nvim_cmd()*
 
     The first argument may also be used instead of count for commands that
-    support it in order to make their usage simpler with |vim.cmd()|. For
+    support it to make their usage simpler with |vim.cmd()|. For
```
+1

justinmk commented on runtime/doc/api.txt:
```diff
@@ -1863,7 +1864,7 @@ nvim_parse_cmd({str}, {opts})                               *nvim_parse_cmd()*
             • pattern: (string) Filter pattern. Empty string if there is no
               filter.
-            • force: (boolean) Whether filter is inverted or not.
+            • force: (boolean) Whether the filter is inverted or not.
```
This is a good change. The `:help dev-doc` guidance against articles is only for the first part of e.g. parameter descriptions and return-value descriptions where "A ..." and "The ..." tend to add limitless noise to the descriptions.

clason commented:
I would recommend to start with what was actually requested: Adding examples or extending explanations in API documentations. This would add much more value for smaller churn.

justinmk commented on runtime/doc/api.txt:
```diff
@@ -1884,7 +1885,7 @@ nvim_parse_cmd({str}, {opts})                               *nvim_parse_cmd()*
           • vertical: (boolean) |:vertical|.
           • split: (string) Split modifier string, is an empty string when
-            there's no split modifier. If there is a split modifier it can be
+            there's no split modifier. If there is a split modifier, it can be
```
Is the grammar tool on "German mode"? Commas are not so commonly needed in English, or at least US English.

clason replied:
It's certainly not in "API reference mode". You should distinguish the parts that are (concise) prose, e.g., in top-level module descriptions and examples, and those that are simply lists for reference. The latter needs to be as terse as possible for layout reasons (good layout is also important for good documentation!), while the former have more room for grammatical changes _if_ they make the text less ambiguous.

justinmk replied:
Both ways are correct, but the docs are currently written in "comma free" style and it's a stylistic choice.

SeniorMars replied:
If needed, I can remove the commas I added, but I feel like it just adds clarity.

justinmk commented on runtime/doc/api.txt:
```diff
@@ -1906,8 +1907,8 @@ nvim_buf_get_option({buffer}, {name})                  *nvim_buf_get_option()*
 nvim_buf_set_option({buffer}, {name}, {value})         *nvim_buf_set_option()*
-    Sets a buffer option value. Passing `nil` as value deletes the option
-    (only works if there's a global fallback)
+    Sets a buffer option value. Passing `nil` as the argument for value deletes 
```
The new wording is more confusing and unnecessarily verbose. "The argument for value" is redundant, because value is an argument.

justinmk commented on runtime/doc/api.txt:
```diff
@@ -1933,9 +1934,9 @@ nvim_get_option({name})                                    *nvim_get_option()*
 
 nvim_get_option_info({name})                          *nvim_get_option_info()*
-    Gets the option information for one option
+    Gets the information of an option
```
good

justinmk commented on runtime/doc/api.txt:
```diff
@@ -1933,9 +1934,9 @@ nvim_get_option({name})                                    *nvim_get_option()*
+    Gets the information of an option
 
-    Resulting dictionary has keys:
+    The resulting dictionary has the following keys:
```
nope. Don't want "the following" everywhere, it is obvious and clear the way it was, with reduced verbosity. This lets the reader focus on the important parts.

SeniorMars commented:
Thanks for commenting! I'll be working on api.txt this week. I used LanguageTool with picky rules. I'll disable that, and I would appreciate your feedback later on as well!

justinmk commented:
Some of these changes look good.

> to have Neovim's documentation similar to Vim's.

I don't think Vim's docs were ran through a grammar tool.

justinmk commented on runtime/doc/api.txt:
```diff
@@ -3413,7 +3415,7 @@ nvim_get_autocmds({*opts})                               *nvim_get_autocmds()*
         • pattern (string): the autocommand pattern. If the autocommand is
           buffer local |autocmd-buffer-local|:
-        • buflocal (boolean): true if the autocommand is buffer local.
+        • buflocal (boolean): true if the autocommand is a local buffer.
```
yes, hyphens are helpful

clason commented on runtime/doc/api.txt:
```diff
@@ -3083,11 +3084,11 @@ nvim_open_win({buffer}, {enter}, {*config})                  *nvim_open_win()*
+                    default highlight group is `FloatTitle`.
+                  • title_pos: Title position must set with the title option.
+                    Value can be of `left` `center` `right` default is left.
+                  • noautocmd: If true, then no buffer-related autocommand
```
Those changes are good.

SeniorMars commented:
> Some of these changes look good.
> 
> > to have Neovim's documentation similar to Vim's.
> 
> I don't think Vim's docs were ran through a grammar tool.

To be clear, I would like to add examples, side effects, and tutorials for using lua inside neovim documentation. 

When I see Vim's docs, I feel I could stay inside vim and use `:help` to customize everything without online help.

clason commented:
> To be clear, I would like to add examples, documentation of side effects, and tutorials for using lua inside neovim documentation.

Be aware that Vim actually has _two_ manuals: The reference manual and the user manual (`usr_xx.txt`). Neovim right now only has a reference manual, which is mostly autogenerated from the docstrings, but no user manual yet at all -- although we have made a first step with `lua-guide.txt`. So tutorials should not be part of, e.g., the api-documentation but be added as new files.

Adding missing examples and important information to the reference manual is definitely welcome (and a better use of everyone's time than messing with punctuation ;))

SeniorMars commented:
I'll then focus on missing examples and dealing with all the TODOs. I'll be back.

I hope to be more active in giving back to this community. Thanks for all your hard work!

SeniorMars committed:
Updated options.c documentation with examples

In addition to grammar, I also added examples I thought were relevant.

SeniorMars committed:
Added spacing and lua tag to examples

SeniorMars committed:
Added @note where it is needed

SeniorMars committed:
Added examples to nearly all api functions
I tried my best to introduce examples with practical use cases, and I
did this by actively looking at examples from plugins/libraries on
github.

Additionally, I added a few commas..., which can introduce different
pose.

SeniorMars committed:
reverted some changes

SeniorMars commented:
Before I continue, I would like feedback on the examples I added. If they are shorter or not to the point, then I can change them.

SeniorMars committed:
Merge branch 'docs' of github.com:SeniorMars/neovim into docs

SeniorMars committed:
docs: Added documentation examples to window.c

Also, corrected a few grammar issues.

SeniorMars committed:
fixup: changed variable names for examples

Note, I'm using my best judgement for variable names.

SeniorMars committed:
fixup: reformatted documentation examples style

SeniorMars committed:
docs: added examples to autocmd.c

In addition, I fixed some grammar mistakes and used @note when needed.

SeniorMars committed:
docs: added examples for tabpage.c

SeniorMars committed:
docs: added examples to win_config.c

Additionally, fixed grammar and spelling mistakes.

SeniorMars committed:
docs: added examples to api/vimscript.c

dundargoc commented:
This PR is too big. Is there a reason why the grammar improvements and example improvements can't be done in a separate pull requests?
